File: Stillinger_Weber.py
import numpy as np

# Given parameters
a = 1.80 + 4
lambda_ = 21.0
alpha = 3.4723e-12
gamma = 1.20 

# Generalized function to calculate h
def h(r1, r2, theta):
    if r1 < a and r2 < a:
        cos_term = np.cos(theta) + 1.0 / 3.0
        exp_term = np.exp(gamma * (1.0 / (r1 - a) + 1.0 / (r2 - a)))
        return lambda_ * exp_term * (cos_term ** 2)
    else:
        return 0

# Function to calculate f3(ri, rj, rk)
def f3(distances, angles):
    #print(f"Debug: distances = {distances}")
    rij, rik, rjk = distances
    theta_jik, theta_ijk, theta_ikj = angles
    
    h1 = h(rij, rik, theta_jik)
    h2 = h(rij, rjk, theta_ijk)
    h3 = h(rik, rjk, theta_ikj)

    return h1 + h2 + h3

File: C++/Stillinger_Weber.cpp
#include <cmath>
#include <array>
using namespace std;

// Constants
const double a = 1.80;
const double lambda_ = 21.0;
const double alpha = 3.4723e-12;
const double gamma = 1.20;

// Function h
double h(double r1, double r2, double theta) {
    if (r1 < a && r2 < a) {
        double cos_term = cos(theta) + 1.0 / 3.0;
        double exp_term = exp(gamma * (1.0 / (r1 - a) + 1.0 / (r2 - a)));
        return lambda_ * exp_term * pow(cos_term, 2);
    } else {
        return 0.0;
    }
}

// Function f3
double f3(const array<double, 3>& distances, const array<double, 3>& angles) {
    double rij = distances[0], rik = distances[1], rjk = distances[2];
    double theta_jik = angles[0], theta_ijk = angles[1], theta_ikj = angles[2];

    double h1 = h(rij, rik, theta_jik);
    double h2 = h(rij, rjk, theta_ijk);
    double h3 = h(rik, rjk, theta_ikj);

    return h1 + h2 + h3;
}
